### ISSUE TITLE
[question] Propagate notifications to the root object

### ISSUE THREAD

**Issue by ewoudwerkman** [pyecore/pyecore#100]
Hi,

Do you know if it is possible to listen to all changes in a resource, e.g. to receive all notifications generated by changing features in an object?

I know I can use `EObserver` to get notification from a specific object in the tree, but I want to receive all notifications.
What approaches are possible here?
I can add an EObserver to all objects in the tree, but there might be other ways to achieve this?

Kind regards,
Ewoud

**Comment by aranega** [comment]
Hi @ewoudwerkman ,

That's a really good question. At the moment, there is no way of doing that beside registering a global listener for each objects in the tree, but I understand that it could be tedious by hand. Another solution would be to register a listener each time an object is inserted in the tree, that could work, it would require to register a "self propagating" observer on the first object you put in a resource, but perhaps there is corner cases. The other solution would be to add a listener list to each resource and to ask each object to notify the resource listeners if they are in a resource. That sounds like the most flexible solution, but I'm not sure about the performance impact, I will measure that asap!

Thanks for the proposal and sorry I cannot give you a more precise solution right now, I will work on that.

Vince

**Comment by ewoudwerkman** [comment]
Hi Vince,

I like the last option. I still have to dive into EMF to find out how they do it there. My current solution is a bit of a hack:

```python
observer = EObserver(notifyChanged=lambda x: handleNotification(command_stack, x))
old_init = EObject.__init__

def new_init(self, **kwargs):
    observer.observe(self)
    old_init(self, **kwargs)

setattr(EObject, '__init__', new_init)
```
but this approach will observe any change to any object (part of a resource or not). I do check in the `handleNotification()` method if the EObject has a resource, and, if not, it will be ignored. As my application is a multi-threaded one, I still have to fix the issue from which resource the change is coming, and if I am interested in that notification.

BTW: my use case is undo/redo for a resource (or resourceSet) and I am extensively using the `commands` package you created: I capture notifications of changes of EObjects and turn them automatically into a set of (Compound)commands to create a undo/redo function in my application.

**Comment by aranega** [comment]
Ok, I have a solution, the thing seems elegant, but at the same time, perhaps overkill... I don't know so much to be honest. I simply said that each `EObject` notify listeners registered on the resource if it is owned by one. This is simple and you can do also the routine of checking the resource at this moment.

The problem with this solution is that there is no notification when an object is set as root of a resource. The question also is "do you need this notification?" and if yes, how to code it? I made some tests where I just said that the `Resource` was actually a `EMetaclass` with a special meta-reference to `EObject` and it works without problem, you can receive `ADD` notification when an object is set as root of a `Resource`, but not when a root element passes from a resource to another. At the same time, I don't know so much if this is what one could want exactly... now that I think about it, it's strange to put the `Resource` as an `EObject`...

With the current solution (without promoting `Resource` as a metaclass):

```python
from pyecore.resources import Resource
from pyecore.notification import EObserver
from pyecore.ecore import *

r1 = Resource()

class GlobalObserve(EObserver):
    def notifyChanged(self, notif):
        print(notif)
        feature = notif.feature
        if feature.is_reference and feature.containment:
            print("My content changed", notif.notifier, "for ref", feature)
            print(notif.new, "has a new parent")
        if feature.is_reference and feature.container:
            print("I am", notif.notifier, "I have a new parent", notif.new, "before", notif.old)

g = GlobalObserve()
g.observe(r1)

A = EClass('A')
r1.append(A)
A.eStructuralFeatures.append(EAttribute("name", EString))
```

which produces:

```
[SET] old=None new=<EClass name="A"> obj=<EAttribute name: EString(str)> #<EReference eContainingClass: <class 'pyecore.ecore.EClass'>>
I am <EAttribute name: EString(str)> I have a new parent <EClass name="A"> before None
[ADD] old=None new=<EAttribute name: EString(str)> obj=<EClass name="A"> #<EReference eStructuralFeatures: <class 'pyecore.ecore.EStructuralFeature'>>
My content changed <EClass name="A"> for ref <EReference eStructuralFeatures: <class 'pyecore.ecore.EStructuralFeature'>>
<EAttribute name: EString(str)> has a new parent
```

Your use-case using all the notifications to automatically build the sequence of commands is really interesting. Do you use this to synchronize remote resources also (it could work)? In the company I was before, we faced someting similar (building commands while giving the impression to deal with real objects). I implemented a kind of transaction system in which you manipulate kind of mirror of your real objects (or full fake one for non existing objects) that records any modifications you ask on the object and gives you the impression that you actually modified the object if you query them. Then, at any moment, you can ask for the CompoundCommand that actually performs all the modification you asked for, and you can play them if you want.

In any case, I'm happy to see that the `command` system works for you, I hesitated a lot before implementing it.

**Comment by ewoudwerkman** [comment]
What changes did you make on `EObject` (or `Resource`) to make this work? 

Having the changes propagated to the Resource seems a very convenient approach. It's not a problem that I don't get the added root resource notification, as this is part of the setup. E.g. when loading a resource, I don't need to record the changes to the tree, only when a loaded resource is changed. The root will in the current use cases always stay the same. My adapted `CommandStack` code allows you to enable and disable recording of changes and record single changes or compound changes.

My current use case is undo/redo in a web application. The backend runs pyecore and flask and hold the model in memory. The web application sends messages to the backend that manipulate the model. These are all kinds of messages: some just change an attribute, others create complete subtrees of the model. As these changes are spread over the backend code, I wanted a single place where these changes are recorded to implement undo/redo. Such that it works for any backend programmer that adapts the model in the backend programatically.

Your mentioned use case is also of interest. Currently our models are stored in a CDO-database (with a REST API), but in order to process them in the frontend, we serialize the model to XML from the database to the backend and in the backend back from XML to pyecore object. This is not efficient (but for historical reasons effective), especially when working with larger models. Recording changes to fake or partial objects and then push those changes to the database seems a beter approach. This recording mechanism might enable this. But the changes to the architecture are quite large, so this will take some time... ;-)

Other use cases we're now looking at are more domain specific. Our ecore-model allows you to specify Energy systems (e.g. the national gas network, a local electricity network, a heat network of a municipality). End users use our toolset to define target energy systems for e.g. 2030 or 2050, that are adaptations (model variations) of the current actual situation. Currently, for each scenario there is a single model instance that describes this. As defining these scenarios is an iterative approach there are multiple versions of each scenario. This gives you easily a dozen model instances for an energy system analysis, that is a big hassle to manage. We're currently looking into questions like: What are the differences between these models? Can I propagate an attribute change from the base scenario to all derived future scenarios, etc. These questions are geared towards modelling differences and applying recorded changes to get a target model instance. This is the reason I'm experimenting with your notifications and commands ;-)

BTW: we open sourced our technology stack recently. See [here](https://github.com/ESDLMapEditorESSIM/docker-toolsuite) for the code. The ecore-model itself was already [available](https://github.com/EnergyTransition/ESDL) (to help adoption of this approach). And documented [here](https://energytransition.gitbook.io/esdl/). What might also be of interest is our [ecore model documentation generator](https://energytransition.github.io/) that generates clickable PlantUML images from the ecore model.

**Comment by aranega** [comment]
I pushed the implementation in the `develop` branch. Basically, it was just to add listeners on the resource and to add them to the elements to notify if the eobject is actually in a resource.

Ok, I totally get your use case. The implementation I provided, manipulating "non-existing" objects was also to deal with undo/redo and sync with a backend. That's nice to see that things seems to integrate quite well (I hope). Asking for changes this way over a model was also our scenario, more or less. Recently, I saw there was the EMF-cloud project that tries to provide a unified API to handle models using json commands and providing different backends. As far as I tested some months ago, there was still some flaws, but I should take a new look to it, perhaps it can be useful for you also?

Regarding the differencies and version between model, we also had this problematic (I was working at [GenMyModel](https://www.genmymodel.com) before going back to academic research). To do a diff, actually, if you already work with commands, you almost have everything, you have the delta of all commands which are required to pass from one version to another. You can work with the sequence of commands between two versions to find which are the elements that have been modified, the one who never changed and check if you can apply the change to your other version (e.g: verifying you didn't deleted an object that should be involved in the modification by side-effect). It could even be a nice research article to write about this.

The ecore model documentation generator looks awesome! Did you release the source code somewhere? I really find very nice the plantUML output, and especially the fact that it's only a sub-part of the whole metamodel which is rendered.

I'm really happy to see that PyEcore is a small brick of a huge and impressive software/system as the one you developed!

**Comment by ewoudwerkman** [comment]
Thanks for the update! I'll try that out.

PyEcore might be a small brick in the system, but it is a very important pillar! By pulling Ecore out of the Eclipse and Java environment and making it available in python, it is far more usable nowadays. I work with many scientists that don't have a software engineering background, but know a little matlab and python and are able to work with this. You lowered the bar to use this modeling framework. I'm happy that you created this!

I learned that if you want a DSL to be accepted by other people than the creators, you need to provide tool support. Python/pyecore is much easier to setup and use than an EMF application in Eclipse, making it easier to create tools and extend the eco system around our model. Furthermore, since pyecore is compliant with EMF Ecore, you can use tools from both environments: Modeling the meta-model in Eclipse, processing models in pyecore, etc. 

Our CDO-based storage for example is written in Java, uses CDO, EMF and EMFCompare (to only store the differences of models in revisions), and the web-based frontend uses pyecore in the backend to render a gui based on the information available in the meta-model, including documentation of classes and attributes. The latter now completely replaces the tree-based editor provided by the EMF.Editor generator, we used in the beginning. If we update the ecore model, we only have to regenerate the classes and that's it: direct and full support of the modelled feature in the frontend. I love it ;-)

Regarding the ecore model documentation generator: it is not open source yet. It is based on an eclipse plugin from [Nasdanika](https://nasdanika.org/builds/master/index.html), which is outdated now and removed from the Eclipse market place recently. Happily Github reserves history and I could get the code from 2018 and adapt it to our needs (e.g. doesn't need Eclipse to run, just a jar-application). The last version of that code base went into a different direction than generating documentation. Keep in mind that it is Java-based. I need to clean it up before I can publish it, but need to find time to do that. If you want access to it, I can try to prioritize that!

I've seen the activity regarding EMF-cloud (and also Sirius on the web), but was not usable for us the last time I checked. I like the approach and am eager to find out if it can be used to extend our eco system.

Then a more general question: what is your opinion about MDE currently. I have the impression that it lost some traction, although there are still issues to solve. Among others, the most important for us is (meta-)model evolution. I've seen that the [Paasage project](https://paasage.ercim.eu/) did some work on that, but that requires some severe adaptation before it is usable for us. Currently I have manual steps to execute when we've evolved our meta-model and need to update all model instances in the database.

**Comment by aranega** [comment]
Thanks so much for the kind words, I'm happy PyEcore fits well in your scenario :). I spent many years with EMF, Eclipse and Java, and as many  I think I felt everything was powerful and handy, but really tedious to use because of all the Eclipse ecosystem. The fact of using Python was something that made sense for me at the time as the syntax is, at the end, not so far from OCL and I the duck typing was really something I missed so many times when I was working with badly designed metamodels. All started thanks to @tarelli that contacted me to work on that for their projects. I really had that in mind for some time, but people around discouraged me of coding that (my love for Python is not shared by everyone). I'm really glad he contacted me as it motivated me to finally do it. As you said, I really have the feeling that all those tools and the Eclipse ecosystem is really powerfull, but the way of installing/dealing with all the plugins is often painful, just writing a small function to extract and information in a project is really cumbersome with Java (i.e: loading the dependency, launching Eclipse, trying to make EMF work outside of Eclipse...etc). 

That's really impressive to have CDO worked flawlessly with EMFCompare. From what I remember, EMFCompare was a little bit surprising in the way it was dealing with some model comparison in the past. I read that latter versions are more "logical", but I never tried myself. I had some prototype using existing Python libraries to code an equivalent in Python, but I wasn't so much happy with the result. I guess I have to be more clear with myself of what I would expect from a model diff exactly. That's really cool that you have a web-based tree editor using PyEcore as backend! Again, you probably know my question, "is it open-source"? That's something that would be definitely really interesting to have to be able to quickly build some models outside of Eclipse. I thought about a "text mode" tree editor to do that, but I have to face the truth that I'm aweful to code non command-line/programmatic user interface...

The documentation generation would be really interesting. Currently, using pyecoregen as basis and pymultigen built by @moltob, something could be done, but it requires just a little bit of work to expose something slightly more high level for any kind of metamodel. If you have time to publish the source code, it would be really cool, but don't put it as a priority, I will not be able to work on that soon as I still struggle with a lot of things lately :p. I really thing that its definitely a very important feature/tool however. There is an issue #83 that have been posted to deal with documentation generation using pydoc. I tried to play with, but there definitely something I don't do right and I didn't had more time to investigate more. Having pyecore objects compliant with pydoc and a documentation generator would really be awesome.

Regarding MDE, I don't know so much... I get the feeling, as you said, that it lost a little bit of traction. It stays really used a lot, but, as you mentionned, even if some problems still need love to work, there is no practical or really usable solutions. The last MoDELS program shows a lot of really interesting papers and we can see that a huge MDE+AI tendency emerges, but some problems seems to have lost interest like metamodel evolution, or as MDE for compilation in general. In our university, we are now teaching more MDE as a practical adopted solution, so I might be biased and get the impression it's still largely used even if I don't get the same feeling of frenzy usage as few years back, especially in the Eclipse ecosystem.

Metamodel and model evolution is something that have been tackled in many article, but at the moment, I'm unaware of practical solutions. The solution we applied in the past was to manually create update scripts that either was crossing the whole database to update all the models, or that were applied "on-demand", at the moment the model was open and an old version of the metamodel was identified. Writing the update script was ok for simple evolutions, but for "detructive" evolutions (metaclass removal...etc), it's always really complicated without human intervention. I tried to be a little bit permissive with PyEcore by letting object accepting also non-estructuralfeature attribute, with this idea to be able to easily mix pyecore objects with plain Python objects, but at the end, it just ease the fact of having a model in a transitionnal state between two versions. CDO doesn't provide an update mechanism? I thought it was, but I never tried it.
